HDDS-6456 Update RocksDB Version to 7.0.4

kerneltime opened the pull request:
## What changes were proposed in this pull request?

Upgrade rocksDB to 7.0.3, which has fix for https://github.com/facebook/rocksdb/issues/8617

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6456

## How was this patch tested?

Tested via a manual build but is blocked on https://github.com/facebook/rocksdb/issues/9689

kerneltime commented:
The build will be fixed once https://github.com/facebook/rocksdb/issues/9689 is resolved

adamretter commented:
RocksJava you 7.0.3 has now been published to Maven Central.

kerneltime commented:
Looks like some of the interfaces have changed, I am taking a look.

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java:
```diff
@@ -268,6 +268,12 @@ public void markCommit(byte[] bytes) throws RocksDBException {
      */
   }
 
+  @Override
+  public void markCommitWithTimestamp(byte[] bytes, byte[] bytes1)
+      throws RocksDBException {
```
```suggestion
  public void markCommitWithTimestamp(final byte[] xid, final byte[] ts)
      throws RocksDBException {
    /**
     * There are no use cases yet for this method in Recon. These will be
     * implemented as and when need arises.
     */
```

I'd suggest removing the `@Override` annotation to improve compatiblity when compiling with an older RocksDB JNI just in case.

I tested with 7.0.2 previously in HDDS-6511. Without the annotation it works as well. Should be the same for 7.0.3

adamretter replied:
@kerneltime Just a thought - If you are not already, you might also like to try switching to the ByteBuffer versions of the methods that move bytes around (as opposed to the older `byte[]` based versions). Several people have reported better performance with the newer API

kerneltime replied:
@adamretter thank you, will look into this.

kerneltime commented:
not sure if this is a flaky test 
``` 
Create key                                                            | FAIL |
255 != 0
------------------------------------------------------------------------------
Generate :: Generate data                                             | FAIL |
2 tests, 1 passed, 1 failed
==============================================================================
Output:  /tmp/smoketest/upgrade/result/robot-1.2.0-finalized-11.xml
jstack 7 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_om1_1_OzoneManagerStarter.stack
jstack 7 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_om2_1_OzoneManagerStarter.stack
jstack 8 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_om3_1_OzoneManagerStarter.stack
jstack 7 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_recon_1_ReconServer.stack
jstack 8 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_s3g_1_Gateway.stack
jstack 8 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_scm_1_StorageContainerManagerStarter.stack
Stopping ha_om3_1   ... 
Stopping ha_s3g_1   ... 
Stopping ha_om1_1   ... 
Stopping ha_dn2_1   ... 
Stopping ha_scm_1   ... 
Stopping ha_dn1_1   ... 
Stopping ha_om2_1   ... 
Stopping ha_recon_1 ... 
Stopping ha_dn3_1   ... 
Stopping ha_recon_1 ... done
Stopping ha_s3g_1   ... done
Stopping ha_om1_1   ... done
Stopping ha_om2_1   ... done
Stopping ha_om3_1   ... done
Stopping ha_scm_1   ... done
Stopping ha_dn1_1   ... done
Stopping ha_dn2_1   ... done
Stopping ha_dn3_1   ... done
Removing ha_om3_1   ... 
Removing ha_s3g_1   ... 
Removing ha_om1_1   ... 
Removing ha_dn2_1   ... 
Removing ha_scm_1   ... 
Removing ha_dn1_1   ... 
Removing ha_om2_1   ... 
Removing ha_recon_1 ... 
Removing ha_dn3_1   ... 
Removing ha_dn1_1   ... done
Removing ha_recon_1 ... done
Removing ha_om3_1   ... done
Removing ha_scm_1   ... done
Removing ha_dn2_1   ... done
Removing ha_s3g_1   ... done
Removing ha_om1_1   ... done
Removing ha_om2_1   ... done
Removing ha_dn3_1   ... done
Removing network ha_net
ERROR: Test execution of /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade is FAILED!!!!
 Log:     /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/result/log.html
Report:  /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/result/report.html
ERROR: Test execution of upgrade is FAILED!!!!
Output:  /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/upgrade.xml
removed 'upgrade/result/1.1.0-1.2.0.xml'
removed 'upgrade/result/log.html'
removed 'upgrade/result/report.html'
renamed 'upgrade/result/1.1.0-1.2.0' -> '/home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/upgrade/1.1.0-1.2.0'
Exception in thread "main" java.net.SocketException: Socket closed
	at java.net.PlainSocketImpl.socketAccept(Native Method)
	at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:409)
	at java.net.ServerSocket.implAccept(ServerSocket.java:560)
	at java.net.ServerSocket.accept(ServerSocket.java:528)
	at org.apache.hadoop.test.JacocoServer.main(JacocoServer.java:60)
Log:     /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/log.html
Report:  /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/report.html
``` 
@sodonnel can we retry the test?

smengcl commented:
> not sure if this is a flaky test
> 
> ```
> Create key                                                            | FAIL |
> 255 != 0
> ------------------------------------------------------------------------------
> Generate :: Generate data                                             | FAIL |
> 2 tests, 1 passed, 1 failed
> ==============================================================================
> Output:  /tmp/smoketest/upgrade/result/robot-1.2.0-finalized-11.xml
> jstack 7 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_om1_1_OzoneManagerStarter.stack
> jstack 7 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_om2_1_OzoneManagerStarter.stack
> jstack 8 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_om3_1_OzoneManagerStarter.stack
> jstack 7 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_recon_1_ReconServer.stack
> jstack 8 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_s3g_1_Gateway.stack
> jstack 8 > /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade/1.1.0-1.2.0/result/ha_scm_1_StorageContainerManagerStarter.stack
> Stopping ha_om3_1   ... 
> Stopping ha_s3g_1   ... 
> Stopping ha_om1_1   ... 
> Stopping ha_dn2_1   ... 
> Stopping ha_scm_1   ... 
> Stopping ha_dn1_1   ... 
> Stopping ha_om2_1   ... 
> Stopping ha_recon_1 ... 
> Stopping ha_dn3_1   ... 
> Stopping ha_recon_1 ... done
> Stopping ha_s3g_1   ... done
> Stopping ha_om1_1   ... done
> Stopping ha_om2_1   ... done
> Stopping ha_om3_1   ... done
> Stopping ha_scm_1   ... done
> Stopping ha_dn1_1   ... done
> Stopping ha_dn2_1   ... done
> Stopping ha_dn3_1   ... done
> Removing ha_om3_1   ... 
> Removing ha_s3g_1   ... 
> Removing ha_om1_1   ... 
> Removing ha_dn2_1   ... 
> Removing ha_scm_1   ... 
> Removing ha_dn1_1   ... 
> Removing ha_om2_1   ... 
> Removing ha_recon_1 ... 
> Removing ha_dn3_1   ... 
> Removing ha_dn1_1   ... done
> Removing ha_recon_1 ... done
> Removing ha_om3_1   ... done
> Removing ha_scm_1   ... done
> Removing ha_dn2_1   ... done
> Removing ha_s3g_1   ... done
> Removing ha_om1_1   ... done
> Removing ha_om2_1   ... done
> Removing ha_dn3_1   ... done
> Removing network ha_net
> ERROR: Test execution of /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/upgrades/non-rolling-upgrade is FAILED!!!!
>  Log:     /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/result/log.html
> Report:  /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/result/report.html
> ERROR: Test execution of upgrade is FAILED!!!!
> Output:  /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/upgrade.xml
> removed 'upgrade/result/1.1.0-1.2.0.xml'
> removed 'upgrade/result/log.html'
> removed 'upgrade/result/report.html'
> renamed 'upgrade/result/1.1.0-1.2.0' -> '/home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/upgrade/1.1.0-1.2.0'
> Exception in thread "main" java.net.SocketException: Socket closed
> 	at java.net.PlainSocketImpl.socketAccept(Native Method)
> 	at java.net.AbstractPlainSocketImpl.accept(AbstractPlainSocketImpl.java:409)
> 	at java.net.ServerSocket.implAccept(ServerSocket.java:560)
> 	at java.net.ServerSocket.accept(ServerSocket.java:528)
> 	at org.apache.hadoop.test.JacocoServer.main(JacocoServer.java:60)
> Log:     /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/log.html
> Report:  /home/runner/work/ozone/ozone/hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/result/report.html
> ```
> 
> @sodonnel can we retry the test?

Yup this is a known flaky test in the upgrade test suite. The root cause being SCM sometimes not being ready when the upgrade test is creating a key (generate phase)

```
15:11:05.927	INFO	Running command 'ozone sh key put /new2-volume/new2-bucket/new2-key /opt/hadoop/NOTICE.txt 2>&1'.	
15:11:08.856	INFO	${rc} = 255	
15:11:08.856	INFO	${output} = INTERNAL_ERROR Allocated 0 blocks. Requested 1 blocks	
```

Retriggered acceptance (misc).

smengcl commented on pom.xml:
```diff
@@ -1636,7 +1636,7 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xs
         <groupId>org.rocksdb</groupId>
         <artifactId>rocksdbjni</artifactId>
-        <version>6.25.3</version>
+        <version>7.0.3</version>
```
```suggestion
        <version>7.0.4</version>
```

Reading through the [7.0.4 changelog](https://github.com/facebook/rocksdb/releases/tag/v7.0.4). We probably should bump this to 7.0.4, once the [mvn repository](https://mvnrepository.com/artifact/org.rocksdb/rocksdbjni) is updated.

kerneltime replied:
Yeah it has not show up on MVN yet.

adamretter replied:
I am working towards 6.29.5 and 7.0.4 releases at the moment, but it will take a little more time...

smengcl commented:
Merged. Thanks @kerneltime for the patch. Thanks @adamretter for publishing the artifact to Maven Central.
